backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/lower/EnumWhenLowering.kt

```diff
@@ -130,7 +54,27 @@ internal class EnumWhenLowering(
         return enumClass.kind == ClassKind.ENUM_CLASS
     }
 
-    private fun createOrdinalVariable(irBlock: IrBlock): IrVariable {
+    override fun visitBlock(expression: IrBlock): IrExpression {
+        if (!shouldLower(expression)) {
+            return super.visitBlock(expression)
+        }
+        val subject = expression.statements[0] as IrVariable
+        // Will be initialized only when we found a branch that compares
+        // subject with compile-time known enum entry.
+        val subjectOrdinalProvider = lazy {
+            createEnumOrdinalVariable(expression)
+        }
+        expression.transformChildrenVoid(EnumEqEqTransformer(context, subject, subjectOrdinalProvider))
```

Comment:
It makes some sense to merge or integrate these two transformers to prevent the entire lowering from taking `O(n^2)` time.
Another option is to cut the recursion in `EnumEqEqTransformer` at blocks with `shouldLower(expression) == true`.